test/test-timer-device.js

```diff
@@ -0,0 +1,218 @@
+  let lastPolled = 0;
+  const poll = curryPollFn(fakeSO, deviceState);
+  t.notOk(poll(1));
+  lastPolled = 1;
+  const cb = makeCallback();
+  const repeaterBuilder = curryRepeaterBuilder(deviceState, () => lastPolled);
+  const rptr = repeaterBuilder(5, 3);
+  rptr.schedule(cb);
+  t.ok(poll(5));
+  lastPolled = 5;
+  t.equals(cb.getCalls(), 1);
+  rptr.schedule(cb);
+  t.notOk(poll(7));
+  lastPolled = 7;
+  t.ok(poll(8));
+  lastPolled = 8;
+  t.equals(cb.getCalls(), 2);
+  t.deepEquals(cb.getArgs(), [rptr, rptr]);
+  t.end();
+});
+
+test('Timer schedule multiple events', t => {
+  const deviceState = buildTimerMap();
+  let lastPolled = 0;
+  const poll = curryPollFn(fakeSO, deviceState);
+  t.notOk(poll(1));
+  lastPolled = 1;
+  const repeaterBuilder = curryRepeaterBuilder(deviceState, () => lastPolled);
+  // will schedule at 2, 5, 8, etc.
+  const rptr = repeaterBuilder(1, 3);
+  const rptrCb = makeCallback();
+  const cb1 = makeCallback();
+  const cb2 = makeCallback();
+  rptr.schedule(rptrCb);
+  poll(4);
+  rptr.schedule(rptrCb);
+  deviceState.add(5, cb1);
+  deviceState.add(6, cb2);
+  poll(7);
+  t.equals(rptrCb.getCalls(), 2);
+  t.deepEquals(rptrCb.getArgs(), [rptr, rptr]);
+  t.equals(cb1.getCalls(), 1);
+  t.equals(cb1.getArgs()[0], null);
+  t.equals(cb2.getCalls(), 1);
+  t.equals(cb2.getArgs()[0], null);
+  t.end();
+});
+
+function makeThrowingCallback() {
+  let threw = false;
+  return {
+    didThrow() {
+      return threw;
+    },
+    wake(_) {
+      threw = true;
+      throw new Error("That didn't work.");
+    },
+  };
+}
+
+test('Timer invoke other events when one throws', t => {
+  const deviceState = buildTimerMap();
+  let lastPolled = 0;
+  const poll = curryPollFn(fakeSO, deviceState);
+  t.notOk(poll(1));
+  lastPolled = 1;
+  const repeaterBuilder = curryRepeaterBuilder(deviceState, () => lastPolled);
+  // will schedule at 2, 5, 8, etc.
+  const rptr = repeaterBuilder(1, 3);
+  const rptrCb = makeCallback();
+  const cb1 = makeThrowingCallback();
+  rptr.schedule(rptrCb);
+  t.ok(poll(4));
+  lastPolled = 4;
+  rptr.schedule(rptrCb);
+  deviceState.add(5, cb1);
+  t.ok(poll(7));
+  lastPolled = 7;
+  t.equals(rptrCb.getCalls(), 2);
+  t.deepEquals(rptrCb.getArgs(), [rptr, rptr]);
+  t.true(cb1.didThrow());
+  t.end();
+});
+
+test('Timer resets on throw', t => {
+  const deviceState = buildTimerMap();
+  let lastPolled = 0;
+  const poll = curryPollFn(fakeSO, deviceState);
+  t.notOk(poll(1));
+  lastPolled = 1;
+  const repeaterBuilder = curryRepeaterBuilder(deviceState, () => lastPolled);
+  // will schedule at 2, 5, 8, etc.
+  const rptr = repeaterBuilder(1, 3);
+  const thrCb = makeThrowingCallback();
+  const cb1 = makeCallback();
+  rptr.schedule(thrCb);
+  t.notOk(poll(4));
+  lastPolled = 4;
+  rptr.schedule(cb1);
+  t.notOk(poll(6));
+  lastPolled = 6;
+  t.equals(cb1.getCalls(), 0);
+  t.deepEquals(cb1.getArgs(), []);
+  t.true(thrCb.didThrow());
+  t.end();
+});
```

Comment:
take a look at test-devices.js for ideas about an integration-style test. We'll need one device, one vat, and then the test code can alternate between calling `poll()` with various times, looking at the return value to make sure the "need to run the queue" flag is correct, then doing `controller.run()` to make sure the callbacks fire.